apis/placement/v1alpha1/stagedupdate_types.go

```diff
@@ -0,0 +1,441 @@
+	// There can be multiple active update runs for each placement but
+	// it's up to the devOps to make sure they don't conflict with each other.
+	// +kubebuilder:validation:Required
+	PlacementRef PlacementReference `json:"placementRef"`
+
+	// The resource snapshot index that this update run is based on.
+	// +kubebuilder:validation:Required
+	ResourceSnapShotIndex string `json:"resourceSnapShotIndex"`
+
+	// The reference to the update strategy that specifies the stages and the sequence
+	// in which the application will be updated on the member clusters. We will compute
+	// the stages according to the referenced strategy when we first start the update run
+	// and record the computed stages in the status field.
+	// +kubebuilder:validation:Required
+	StagedUpdateStrategyRef v1beta1.NamespacedName `json:"stagedRolloutStrategyRef"`
+}
+
+// PlacementReference is a reference to a placement object.
+type PlacementReference struct {
+	// Name is the name of the referenced placement.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// +genclient
+// +genclient:namespaced
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope="Namespaced",categories={fleet,fleet-placement}
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// StagedUpdateStrategy defines a reusable strategy that specifies the stages and the sequence
+// in which the application will be updated on the member clusters.
+type StagedUpdateStrategy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Stage configurations for each update stage.
+	// +kubebuilder:validation:MaxItems=127
+	// +kubebuilder:validation:Required
+	Stages []StageConfig `json:"stages"`
+}
+
+// StagedUpdateStrategyList contains a list of StagedUpdateStrategy.
+// +kubebuilder:resource:scope="Namespaced"
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type StagedUpdateStrategyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []StagedUpdateStrategy `json:"items"`
+}
+
+// StageConfig describes a single update stage group configuration.
+// The clusters in each stage are updated sequentially for now.
+// We will stop the update if any of the updates fail.
+type StageConfig struct {
+	// The name of the stage. This MUST be unique within the same StagedUpdateStrategy.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// LabelSelector is a label query over all the joined member clusters. Clusters matching the query are selected
+	// for this stage. There cannot be overlapping clusters between stages when the stagedUpdateRun is created.
+	// If the label selector is absent, all the selected clusters are selected.
+	// +kubebuilder:validation:Optional
+	LabelSelector *metav1.LabelSelector `json:"labelSelector"`
```

Comment:
We are going to do an intersection between the clusters selected by scheduler and the clusters selected here right?